evalml/pipelines/pipeline_meta.py

```diff
@@ -25,9 +46,8 @@ def _check_for_fit(self, X=None, y=None, objective=None):
             elif y is None:
                 return method(self, X)
             # For time series classification pipelines, predict will take X, y, objective
-            elif len(inspect.getfullargspec(method).args) == 4:
+            elif method.__name__ == 'predict':
```

Comment:
If adding this case solves the problem for time series problems could we just add this case to `PipelineBaseMeta`? I think its better organizationally to use inheritance but since you can't inherit the inner function, the current implementation has some code reuse.

```
@wraps(method)
        def _check_for_fit(self, X=None, y=None, objective=None):
            klass = type(self).__name__
            if not self._is_fitted:
                raise PipelineNotYetFittedError(f'This {klass} is not fitted yet. You must fit {klass} before calling {method.__name__}.')
           elif klass == '...` and method.__name__ == 'predict':
                return method(self, X, y, objective)
            elif X is None and y is None:
                return method(self)
            elif y is None:
                return method(self, X)
            else:
                return method(self, X, y)
```

Reply:
I'm down for either! You're right, maybe since we only have one method right now its not worth inheritance... Took the suggestion to subclass via @freddyaboulton and @dsherry 's discussion on the issue. Do you guys have thoughts about this?

Reply:
@angela97lin I think the apis for time series vs non-time series pipelines are different enough that trying to handle them both in one metaclass makes the implementation pretty tricky to understand/maintain.

I think having the separate metaclasses makes the distinction between time-series vs non-time-series clear and should (hopefully) make the meta class implementations easier to understand.

To that end, this is what I was thinking we'd do when we wrote up the issue:

1. Delete `y` from the `PipelineBaseMeta` `_check_for_fit` helper since `y` is never allowed for non time series pipelines.
2. Rather than switching based on whether `X` and/or `y` are None, switch on the method name like you did for predict `if method.__name__ == 'predict':``elif method.__name__ == 'feature_importance`

I don't have a strong opinion about whether we should inherit from `PipelineBaseMeta` or `BaseMeta` since the work is being done by the inter `_check_for_fit` function which can't be shared.

Let me know what you think!

Reply:
@freddyaboulton Thank you for explaining your thoughts! I like your suggestions a lot.

- I agree with the idea of having them as two separate metaclasses. I agree that it makes it a bit more clear in terms of logic and implementation.
- I deleted `y` from the `PipelineBaseMeta`
- RE using names: I decided to update to check via name (though `method.__name__ == 'feature_importance'` ends up being `method.__name__ == '__get__'` since feature importance is an attribute, so I threw it in the `else` case). I do think for now, this makes things clear because we're only checking a few methods/attributes (predict_proba, predict, feature_importance). I do wonder if this will be maintainable once we add more methods to check, but we might never get there so I'm all for readability! 🥳

Reply:
@angela97lin Awesome! I didn't know that the "method" passed to the meta class would be `__get__` for properties. I think it's ok to catch all properties in the `else` clause because `method(self)` should work for all properties because they don't take arguments!